pkg/adaptor/cloud/libvirt/libvirt.go

```diff
@@ -116,55 +128,196 @@ func uploadIso(isoFile string, isoVolName string, libvirtClient *libvirtClient)
 
 }
 
-func CreateDomain(ctx context.Context, libvirtClient *libvirtClient, v *vmConfig) (result *createDomainOutput, err error) {
+func getGuestForArchType(caps *libvirtxml.Caps, arch string, virttype string) (*libvirtxml.CapsGuest, error) {
+	for _, guest := range caps.Guests {
+		if guest.Arch.Name == arch && guest.OSType == virttype {
+			return &guest, nil
+		}
+	}
+	return nil, fmt.Errorf("could not find any guests for architecture type %s/%s", virttype, arch)
+}
 
-	v.cpu = uint(2)
-	v.mem = uint(8)
-	v.rootDiskSize = uint64(10)
+func getHostCapabilities(conn *libvirt.Connect) (*libvirtxml.Caps, error) {
+	// We should perhaps think of storing this on the connect object
+	// on first call to avoid the back and forth
+	capsXML, err := conn.GetCapabilities()
+	if err != nil {
+		return nil, fmt.Errorf("unable to get capabilities, cause: %w", err)
+	}
 
-	exists, err := checkDomainExistsByName(v.name, libvirtClient)
+	caps := &libvirtxml.Caps{}
+	err = xml.Unmarshal([]byte(capsXML), caps)
 	if err != nil {
-		return nil, fmt.Errorf("Error in checking instance: %s", err)
+		return nil, fmt.Errorf("unable to unmarshal capabilities, cause: %w", err)
 	}
-	if exists {
-		logger.Printf("Instance already exists ")
-		return &createDomainOutput{
-			instance: v,
-		}, nil
+
+	return caps, nil
+}
+
+func lookupMachine(machines []libvirtxml.CapsGuestMachine, targetmachine string) string {
+	for _, machine := range machines {
+		if machine.Name == targetmachine {
+			if machine.Canonical != "" {
+				return machine.Canonical
+			}
+			return machine.Name
+		}
 	}
+	return ""
+}
 
-	rootVolName := v.name + "-root.qcow2"
-	err = createVolume(rootVolName, v.rootDiskSize, libvirtClient.volName, libvirtClient)
+func getCanonicalMachineName(caps *libvirtxml.Caps, arch string, virttype string, targetmachine string) (string, error) {
```

Comment:
@CarstenLeue can you please help me understand this method. Sorry I don't understand what this is supposed to do and how this behaves for different arch.

Reply:
@bpradipt this method returns information about the actual guest machine to start and it selects the canonical (i.e. preferred) machine. This is what's listed in the response to a `virsh capabilities` on a libvirt environment.

For my test case this e.g. returns

```
s390-ccw-virtio-rhel9.0.0
```

Note that this path is only called for s390x. I wanted to avoid to change anything in the x86 path which is as I expect working and tested.

This method could also be invoked for x86 based libvirt installations, if desired.

Reply:
@CarstenLeue thanks. This is helpful. For x86 the defaults gets used so we don't need to explicitly set the `machine canonical` name etc.
I have two additional questions and that should be it for this PR :-) 

1. Since the canonical machine name, guest details (ostype, arch, emulator) is not going to change for the libvirt connection (we are using only one libvirt host for now), would it be better to retrieve these during initialisation and store it in a struct to avoid invoking the methods for every VM creation? 
2. Does it make sense to add a brief description for the new methods with some examples o/p snippets like shown below, to help new developers?
```
 <guest>
    <os_type>hvm</os_type>
    <arch name='x86_64'>
      <wordsize>64</wordsize>
      <emulator>/usr/bin/qemu-system-x86_64</emulator>
```
```
  <machine canonical='pc-i440fx-6.1' maxCpus='255'>pc</machine>
  <machine canonical='pc-q35-6.1' maxCpus='288'>q35</machine>
```

Reply:
Hi @bpradipt 

1. the `getCanonicalMachineName` method does not make any remote call to the host, it is just operating on existing data structures. In priniciple we could cache this piece of information. However since there is no real overhead I prefer to keep these methods in place, because this keeps the code cleanly separated. At least for me its easier to read and understand the code that way, otherwise if the code depends on pre-initialized data as a reader I'll have to worry about understanding how this initialization works. 

If it really turns out to be a performance problem, I'd rather make the implementation memoized (which is possible since it's a pure function).

2. sure, I'll add some comments to the code

Reply:
Thanks @CarstenLeue